motionplan/collision.go

```diff
@@ -49,301 +46,103 @@ func collisionListsAlmostEqual(cs1, cs2 []Collision) bool {
-	reference *CollisionSystem,
-	reportDistances bool,
-) (*collisionGraph, error) {
-	var err error
-	cg := &collisionGraph{
-		key:             key,
-		test:            test,
-		adjacencies:     make([][]float64, key.count()),
-		triangular:      key == test,
-		reportDistances: reportDistances,
+func newCollisionGraph(x, y map[string]spatial.Geometry, reference *collisionGraph, reportDistances bool) (cg *collisionGraph, err error) {
+	if y == nil {
+		y = x
 	}
-	for i := range cg.adjacencies {
-		cg.adjacencies[i] = make([]float64, test.count())
-		keyi := key.entityFromIndex(i)
-		startIndex := 0
-		if cg.triangular {
-			startIndex = i + 1
-			for j := 0; j < startIndex; j++ {
-				cg.adjacencies[i][j] = math.NaN()
+	cg = &collisionGraph{newGeometryGraph(x, y, reportDistances)}
+
+	var distance float64
+	for xName, xGeometry := range cg.x {
+		for yName, yGeometry := range cg.y {
+			if _, ok := cg.distances[yName][xName]; ok || xGeometry == yGeometry {
+				continue
 			}
-		}
-		for j := startIndex; j < len(cg.adjacencies[i]); j++ {
-			testj := test.entityFromIndex(j)
-			if reference.CollisionBetween(keyi.name, testj.name) {
-				cg.adjacencies[i][j] = math.NaN() // represent previously seen collisions as NaNs
-			} else {
-				cg.adjacencies[i][j], err = test.checkCollision(keyi, testj, reportDistances)
-				if err != nil {
-					return nil, err
-				}
-				if !reportDistances && cg.adjacencies[i][j] > -spatial.CollisionBuffer {
-					return cg, nil
-				}
+			if reference != nil && reference.collisionBetween(xName, yName) {
+				distance = math.NaN() // represent previously seen collisions as NaNs
+			} else if distance, err = cg.checkCollision(xGeometry, yGeometry); err != nil {
+				return nil, err
+			}
+			cg.distances[xName][yName] = distance
+			if !reportDistances && distance <= spatial.CollisionBuffer {
+				return cg, nil
 			}
 		}
 	}
 	return cg, nil
 }
 
-func (cg *collisionGraph) getIndices(keyName, testName string) (int, int, bool) {
-	i := cg.key.indexFromName(keyName)
-	j := cg.test.indexFromName(testName)
-	return i, j, i >= 0 && j >= 0
+func (cg *collisionGraph) checkCollision(x, y spatial.Geometry) (float64, error) {
+	if cg.reportDistances {
+		return x.DistanceFrom(y)
+	}
+	col, err := x.CollidesWith(y)
+	if col {
+		return -1, err
+	}
+	return 1, err
 }
 
 // collisionBetween returns a bool describing if the collisionGraph has an edge between the two entities that are specified by name.
-func (cg *collisionGraph) collisionBetween(keyName, testName string) bool {
-	if i, j, ok := cg.getIndices(keyName, testName); ok {
-		if cg.triangular && i > j {
-			i, j = j, i
-		}
-		if cg.adjacencies[i][j] >= -spatial.CollisionBuffer {
-			return true
+func (cg *collisionGraph) collisionBetween(name1, name2 string) bool {
+	distance, ok := cg.distances[name1][name2]
+	if !ok {
+		if distance, ok = cg.distances[name2][name1]; !ok {
+			return false
 		}
 	}
-	return false
+	return distance <= spatial.CollisionBuffer
```

Comment:
What does this do if `distance` is `math.NaN`?

Reply:
Any comparison with `NaN` returns false per IEEE standards, which is what we want in this case